grpc-gcp/src/main/java/com/google/cloud/grpc/multiendpoint/MultiEndpoint.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.grpc.multiendpoint;
+
+import static java.util.Comparator.comparingInt;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+import com.google.cloud.grpc.multiendpoint.Endpoint.EndpointState;
+import com.google.common.base.Preconditions;
+import com.google.errorprone.annotations.CheckReturnValue;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+
+/**
+ * MultiEndpoint holds a list of endpoints, tracks their availability and defines the current
+ * endpoint. An endpoint has a priority defined by its position in the list (first item has top
+ * priority). MultiEndpoint returns top priority endpoint that is available as current. If no
+ * endpoint is available, MultiEndpoint returns the top priority endpoint.
```

Comment:
Yes, but the user of MultiEndpoint does not want to know if any or all endpoints is available. The user just wants the best endpoint possible. If all endpoints are unavailable the user would still want to try to send an RPC (even if it will be queued).

Reply:
grpc-gcp-java does not introduce any new behavior here, it would be exactly the same with vanilla gRPC client.

We don't throw an error right away because the network issue could resolve very soon, if it's not the error will be raised after the timeout.

Reply:
Thanks for the explanation!